### DISCUSSION TITLE
Provider: SSL does not work if SSL_CTX_new is called before OSSL_STORE_open_ex(private key)

### DISCUSSION THREAD

**Discussion by space88man** [openssl/openssl#23727] [Q&A]
When using other provider keys  for SSL  should the sequence of SSL_CTX_new and OSSL_STORE_open_ex(to get the private key) matter?

Use case: 3rd party provider (pkcs11-provider(https://github.com/latchset/pkcs11-provider) with NSS softtoken as the PKCS#11 module) where the private key is from a PKCS#11 module.

Apps like `openssl req -key "pkcs11:type=private;object=somekey"` `openssl pkey -in  "pkcs11:type=private;object=somekey"` all work.

In my code I found that if `SSL_CTX_new` is called before `OSSL_STORE_open_ex` then 
- the private key is found and `SSL_CTX_check_private_key` works
- SSL fails

During SSL_accept no calls are forwarded to the provider.

If the private key is found using OSSL_STORE_open_ex first then SSL_CTX_new then it works (this is the order that s_server uses so s_server actually works)

It is as if the provider doesn't hook into the SSL state machine if it is called after SSL_CTX_new.

Is there a more explicit way to get a provider to be hooked into SSL (the sequencing of SSL_CTX_new and OSSL_STORE_open_ex is fragile)

The code is very simple: sslecho/main.c (from demos) adjust to load private keys from provider.

A reproducer can be found here: https://github.com/latchset/pkcs11-provider/files/14461489/example-fail.zip

**Comment by ZPedro** [comment]
Can confirm: when we added support for that provider, including the PKCS#11 URI, to our software that integrates OpenSSL to serve as a server, we spontaneously went with that order of API calls; we more specifically got this error:

`167772278 no suitable signature algorithm`

as a result of the SSL_accept() call.

Sequencing to call OSSL_STORE_open_ex() before SSL_CTX_new_ex() gets us past this hurdle: a common chipersuite does get selected.

**Reply by ZPedro** [reply]
Update: we have decided to update pkcs11-provider to v1.0 (by building it from source) in order to avoid having to figure out and maintain such fragile workarounds: even after we cleared the previously mentioned hurdle, we had gotten stuck on another error code returned by SSL_accept() and there did not seem to be a readily available workaround to clear that further hurdle. Our use of this v1.0 was successful: we no longer observe any OpenSSL error from SSL_accept() and our TLS server scenario completes, we were even able to remove the workaround mentioned here.

For context, at the moment we rely on the OpenSSL 3.0.13 version included with Ubuntu 24.04, and we were relying on the pkcs11-provider 0.3-1build2 version also included with Ubuntu 24.04.
